idseq_pipeline/commands/curate_accession2taxid.py

```diff
@@ -0,0 +1,143 @@
+        dest_dir = os.path.join(DEST_DIR, 'accession2taxid')
+        execute_command("mkdir -p %s" % dest_dir)
+
+        # Retrieve the reference files
+        print "Retrieving references"
+        arguments = self.options
+        threads = {}
+        threads['nt'] = MyThread(target=download_reference_locally_with_version_any_source_type,
+            args=(arguments['--nt_file'], dest_dir, dest_dir))
+        threads['nt'].start()
+        threads['nr'] = MyThread(download_reference_locally_with_version_any_source_type,
+            (arguments['--nr_file'], dest_dir, dest_dir))
+        threads['nr'].start()
+        mapping_files_local = []
+        mapping_version_numbers = []
+        mapping_files_sources = arguments['--mapping_files'].split(",")
+        mapping_file_results = {}
+        for f in mapping_files_sources:
+            assert f not in threads
+            threads[f] = MyThread(target=download_reference_locally_with_version_any_source_type,
+                args=(f, dest_dir, dest_dir))
+            threads[f].start()
+        for f in threads:
+            threads[f].join()
+            assert not threads[f].exception
+        nt_file_local, nt_version_number = threads['nt'].result
+        nr_file_local, nr_version_number = threads['nr'].result
+        for f in mapping_files_sources:
+            mapping_file_local, mapping_version_number = threads[f].result
+            mapping_files_local.append(mapping_file_local)
+            mapping_version_numbers.append(mapping_version_number) 
+        print "Reference download finished"
+
+        # Produce the output file
+        print "Curating accession2taxid"
+        output_mapping_file = os.path.join(DEST_DIR, 'curated_accession2taxid.txt')
+        curate_taxon_dict(nt_file_local, nr_file_local, mapping_files_local, output_mapping_file)
+        print "Curation finished"
+
+        # Convert to a berkeley db
+        print "Writing berkeley db"
+        output_db_file = os.path.join(dest_dir, 'curated_accession2taxid.db')
+        execute_command("rm -f %s" % output_db_file)
+        generate_accession2taxid_db(output_mapping_file, output_db_file, False)
+        output_s3_path = arguments['--output_s3_folder'].rstrip('/')
+        output_s3_file_gz = os.path.join(output_s3_path, os.path.basename(output_db_file) + ".gz")
+
+        # Record difference with old accession2taxid
+        previous_mapping_s3 = arguments.get('--previous_mapping')
+        if previous_mapping_s3:
+            print "Computing diff"
+            previous_mapping_local = os.path.join(dest_dir, "previous_mapping.db.gz")
+            execute_command("aws s3 cp --quiet %s %s" % (previous_mapping_s3, previous_mapping_local))
+            execute_command("gunzip %s" % previous_mapping_local)
+            previous_mapping = shelve.open(os.path.splitext(previous_mapping_local)[0])
+            new_mapping = shelve.open(output_db_file)
+            previous_accessionids = set(previous_mapping.keys())
```

Comment:
don't you need gigantic memory for this if each set? You can also loop through the new ones and see what's missing in the old one.

Reply:
Ok, I'll submit another diff to convert to loop. To be clear: opening a python shelf and writing to it doesn't keep the entire map in memory right? It writes to the file while keeping very little in memory? Because otherwise we would need very large memory in the other steps anyway to produce the shelves.